File: python/178.py
"""Project Euler Problem 178: Step Numbers."""

MAX_LEN = 40
DIGIT_COUNT = 10
FULL_MASK = (1 << DIGIT_COUNT) - 1


def main() -> int:
    """Main function."""
    # dp[digit][mask] -> number of step numbers of current length
    current: List[List[int]] = [[0] * (1 << DIGIT_COUNT) for _ in range(DIGIT_COUNT)]

    for digit in range(1, 10):
        current[digit][1 << digit] = 1

    total = 0

    for step in range(MAX_LEN - 1):
        next_dp: List[List[int]] = [[0] * (1 << DIGIT_COUNT) for _ in range(DIGIT_COUNT)]
        for digit in range(10):
            mask_counts = current[digit]
            for mask in range(FULL_MASK + 1):
                count = mask_counts[mask]
                if count == 0:
                    continue

                if digit > 0:
                    next_digit = digit - 1
                    new_mask = mask | (1 << next_digit)
                    next_dp[next_digit][new_mask] += count

                if digit < 9:
                    next_digit = digit + 1
                    new_mask = mask | (1 << next_digit)
                    next_dp[next_digit][new_mask] += count

        current = next_dp

        # After step k, we have numbers of length k+2
        # We want to count numbers of length >= DIGIT_COUNT
        if step + 2 >= DIGIT_COUNT:
            for digit in range(10):
                total += current[digit][FULL_MASK]

    return total


if __name__ == "__main__":
    print(main())

File: c/178.c
/*
 * Project Euler Problem 178: Step Numbers
 *
 * Count step numbers of length 10..40 that contain all digits 0-9.
 * A step number has consecutive digits differing by 1.
 * DP: dp[digit][mask] = count of step numbers ending in digit with
 * digit-set mask.
 */
#include <stdio.h>
#include <string.h>

#define MAX_LEN 40
#define FULL_MASK 1023  /* (1 << 10) - 1 */

int main(void) {
    long long cur[10][1024];
    long long nxt[10][1024];

    memset(cur, 0, sizeof(cur));

    /* Initialize: 1-digit step numbers (digits 1-9) */
    for (int d = 1; d <= 9; d++)
        cur[d][1 << d] = 1;

    long long total = 0;

    for (int step = 0; step < MAX_LEN - 1; step++) {
        memset(nxt, 0, sizeof(nxt));

        for (int digit = 0; digit <= 9; digit++) {
            for (int mask = 0; mask <= FULL_MASK; mask++) {
                long long count = cur[digit][mask];
                if (count == 0) continue;

                if (digit > 0) {
                    int nd = digit - 1;
                    int nm = mask | (1 << nd);
                    nxt[nd][nm] += count;
                }
                if (digit < 9) {
                    int nd = digit + 1;
                    int nm = mask | (1 << nd);
                    nxt[nd][nm] += count;
                }
            }
        }

        memcpy(cur, nxt, sizeof(cur));

        /* After step k, we have numbers of length k+2 */
        if (step + 2 >= 10) {
            for (int digit = 0; digit <= 9; digit++)
                total += cur[digit][FULL_MASK];
        }
    }

    printf("%lld\n", total);
    return 0;
}
